cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java

```diff
@@ -0,0 +1,240 @@
+        // eventually cache entries will expire and there won't be stale privileges
+        LOG.debug("Error while updating privileges for {}", principal, e);
+        LOG.warn("Error while updating privileges for {}.", principal);
+      }
+    }
+  }
+
+  @Override
+  protected ScheduledExecutorService executor() {
+    executor = Executors.newSingleThreadScheduledExecutor(
+      Threads.createDaemonThreadFactory(String.format("authorization-%s-service", serviceName)));
+    return executor;
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.debug("Shutting down authorization {} service...", serviceName);
+    authPolicyCache.invalidateAll();
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+    LOG.info("Shutdown authorization {} service successfully.", serviceName);
+  }
+
+  @VisibleForTesting
+  Map<Principal, Map<EntityId, Set<Action>>> getCache() {
+    return authPolicyCache.asMap();
+  }
+
+  protected Map<EntityId, Set<Action>> fetchPrivileges(Principal principal) throws Exception {
+    State serviceState = state();
+    // The only states in which the service can be used are:
+    // 1. STARTING - while pre-populating the cache with the current user's privileges
+    // 2. RUNNING - from calls to enforce() during program execution, after the service has been started successfully
+    // 3. STOPPING - while invalidating the cache during service stop.
+    if (!SERVICE_AVAILABLE_STATES.contains(serviceState)) {
+      throw new IllegalStateException(
+        String.format("Cannot use authorization %s service because it has not been started. Its current state is %s.",
+                      serviceName, serviceState)
+      );
+    }
+    Map<EntityId, Set<Action>> result = new HashMap<>();
+    Set<Privilege> privileges = privilegesFetcher.listPrivileges(principal);
+    if (privileges == null) {
+      return result;
+    }
+
+    for (Privilege privilege : privileges) {
+      Set<Action> actions = result.get(privilege.getEntity());
+      if (actions == null) {
+        actions = EnumSet.noneOf(Action.class);
+        result.put(privilege.getEntity(), actions);
+      }
+      actions.add(privilege.getAction());
+    }
+
+    return result;
+  }
+
+  protected Map<EntityId, Set<Action>> getPrivileges(Principal principal) throws Exception {
+    return cacheEnabled ? authPolicyCache.get(principal) : fetchPrivileges(principal);
+  }
+
+  /**
+   * On an authorization-enabled cluster, if caching is enabled too, updates the cache in the
+   * {@link AuthorizationEnforcementService} with the privileges of the user running the program.
+   */
+  private void updatePrivilegesOfCurrentUser() {
+    String userName;
+    try {
+      userName = UserGroupInformation.getCurrentUser().getShortUserName();
+    } catch (IOException e) {
+      LOG.warn("Error while determining the currently logged in user. Skipping pre-population of authorization cache.",
```

Comment:
This could be a transient error. Its an optimization to load privileges before starting the program. So its just a warning. If it continues to fail after the program starts, it will become fatal.